common/src/main/cpp/classes/openxr_fb_render_model.cpp

```diff
@@ -0,0 +1,138 @@
+	return render_model_type;
+}
+
+bool OpenXRFbRenderModel::has_render_model_node() {
+	return render_model_node != nullptr;
+}
+
+void OpenXRFbRenderModel::load_render_model() {
+	if (render_model_node != nullptr) {
+		render_model_node->queue_free();
+		render_model_node = nullptr;
+	}
+
+	String render_model_path;
+	switch (render_model_type) {
+		case MODEL_CONTROLLER_LEFT: {
+			render_model_path = "/model_fb/controller/left";
+		} break;
+
+		case MODEL_CONTROLLER_RIGHT: {
+			render_model_path = "/model_fb/controller/right";
+		} break;
+
+		default: {
+			render_model_path = "";
+		}
+	}
+
+	if (render_model_path.is_empty()) {
+		return;
+	}
+
+	PackedByteArray render_model_buffer = OpenXRFbRenderModelExtensionWrapper::get_singleton()->get_buffer(render_model_path);
+	if (render_model_buffer.is_empty()) {
+		UtilityFunctions::print_verbose("Failed to load render model buffer from path [", render_model_path, "] in OpenXRFbRenderModel node");
+		return;
+	}
+
+	Ref<GLTFDocument> gltf_document;
+	gltf_document.instantiate();
+	Ref<GLTFState> gltf_state;
+	gltf_state.instantiate();
+
+	Error err = gltf_document->append_from_buffer(render_model_buffer, "", gltf_state);
+	if (err != OK) {
+		UtilityFunctions::print_verbose("Failed to instance render model in OpenXRFbRenderModel node");
+		return;
+	}
+
+	render_model_node = Object::cast_to<Node3D>(gltf_document->generate_scene(gltf_state));
+	if (render_model_node) {
+		add_child(render_model_node);
+		emit_signal("openxr_fb_render_model_loaded");
+	}
+}
+
+Node3D *OpenXRFbRenderModel::get_render_model_node() {
+	return render_model_node;
+}
+
+void OpenXRFbRenderModel::_notification(int p_what) {
+	switch (p_what) {
+		case NOTIFICATION_POSTINITIALIZE: {
+			Ref<OpenXRInterface> openxr_interface = XRServer::get_singleton()->find_interface("OpenXR");
+			if (openxr_interface.is_valid()) {
+				openxr_interface->connect("session_begun", callable_mp(this, &OpenXRFbRenderModel::load_render_model));
+			}
+		} break;
+		case NOTIFICATION_ENTER_TREE: {
+			if (OpenXRFbRenderModelExtensionWrapper::get_singleton()->is_openxr_session_active()) {
+				load_render_model();
+			}
+		} break;
+	}
+}
+
+void OpenXRFbRenderModel::_bind_methods() {
+	ClassDB::bind_method(D_METHOD("has_render_model_node"), &OpenXRFbRenderModel::has_render_model_node);
+	ClassDB::bind_method(D_METHOD("get_render_model_node"), &OpenXRFbRenderModel::get_render_model_node);
+	ClassDB::bind_method(D_METHOD("set_render_model_type", "render_model_type"), &OpenXRFbRenderModel::set_render_model_type);
+	ClassDB::bind_method(D_METHOD("get_render_model_type"), &OpenXRFbRenderModel::get_render_model_type);
+
+	ADD_PROPERTY(PropertyInfo(Variant::INT, "render_model_type", PROPERTY_HINT_ENUM, "Left Controller,Right Controller"), "set_render_model_type", "get_render_model_type");
+
+	BIND_ENUM_CONSTANT(MODEL_CONTROLLER_LEFT);
+	BIND_ENUM_CONSTANT(MODEL_CONTROLLER_RIGHT);
+
+	ADD_SIGNAL(MethodInfo("openxr_fb_render_model_loaded"));
```

Comment:
I'm not sure about the best implementation but the one thing I was missing is control/signals related to the controller no longer being tracked.

Right now in the demo you react to the `openxr_fb_render_model_loaded` signal, and hide the hand tracking but that seems to be too restrictive, and once we use proper hand tracking, it should actually show the hand properly holding the controller (and thus verifying the positioning is correct).

Off course this could be a limitation of the demo, and possibly the logic I'm looking for shouldn't be part of this but should be part of the general hand tracking implementation that is in the core of Godot.

If the latter, we should probably do a follow up PR that adds hand tracking into the demo with the hand meshes Malcolm has (I plan to do a version of this during the week for the `godot-demo-projects` repo).

But if there is more to handle here, especially if tracking of controllers stops and the controllers should be hidden, that is worth looking into before we merge this.

Reply:
My initial thought: this extension exists to provide 3D models, so our implementation should just expose those 3D models, and any further logic should be done by end users (or at least not here). It's possible users may want to drop static controller models in other areas, not just use them for the tracked controllers.

Reply:
I agree with that. There's also more models than just controllers, this can also be used for physics keyboards (for the tracked keyboard feature) and the advanced virtual keyboard (which isn't tracking anything at all).